Feature/faster sss trajectories

LoyVanBeek opened the pull request:
- fixes https://github.com/mojin-robotics/cob4/issues/1277

TODO in this PR:

- [x] Take velocity and acceleration of each joint into account instead of 1 velocity and 1 acceleration for all joints
- [x] Get velocity limits from URDF. <strike>Not sure yet how to get relation between joint name and index in goal position array </strike>
- [x] Allow velocity scaling

LoyVanBeek committed:
Explicitly import numpy. This was implicitly imported via at least tf.transformations and possibly others

LoyVanBeek committed:
Iterate over all joints to calculate_point_time and take the  slowest time for all joints

Earlier only 1 joint velocity and acceleration was used.

TOOD: Read velocity (and acceleration?) from URDF

fmessmer reviewed (dismissed):
looks ok, needs testing on a robot though
@floweisshardt

fmessmer commented:
@LoyVanBeek @floweisshardt 
what about that special hack regarding https://github.com/ipa320/cob_command_tools/blob/indigo_dev/cob_script_server/src/simple_script_server/simple_script_server.py#L621-L622 where it needs to be decided whether the next trajectory point goes in the same direction for each joint (no stopping) or - at least - one joint changes direction (needs stopping)

floweisshardt commented:
> @LoyVanBeek @floweisshardt
> what about that special hack regarding https://github.com/ipa320/cob_command_tools/blob/indigo_dev/cob_script_server/src/simple_script_server/simple_script_server.py#L621-L622 where it needs to be decided whether the next trajectory point goes in the same direction for each joint (no stopping) or - at least - one joint changes direction (needs stopping)

I opened another issue for that: https://github.com/mojin-robotics/cob4/issues/1304

fmessmer commented:
this repo has been released/tagged in the meantime, please rebase

LoyVanBeek commented:
@floweisshardt 
For the speed scaling: what is the API we want for this? I'd say `sss.move('arm', 'home', blocking=True, speed_scale=0.5)`
This then multiplies each joint velocity with the `speed_scale`

fmessmer commented:
call it `speed_factor`...default should be `1.0`

I think we should finalize this pr together with #260 
@LoyVanBeek once you have implemented your features, we should create a merged feature branch and test/finalize on hardware...

LoyVanBeek commented:
Another API design thing:
`simple_script_server` gets a default speed from the parameter server, one speed for all joints whether it s revolute or prismatic etc. 

Getting this from URDF is simple enough, but is likely to be different. Users of the SSS who don't know of this change will get different behavior, with a potentially much faster or slower robot. How to tackle this?

Do we want to use the `default_vel` from the param server for those joints URDF has no velocity for?

LoyVanBeek committed:
Get joint velocity from URDF in compose_trajectory

LoyVanBeek committed:
fixup! Iterate over all joints to calculate_point_time and take the  slowest time for all joints

LoyVanBeek committed:
Extend sss.move with a speed_factor to speed up movements

LoyVanBeek commented:
> I think we should finalize this pr together with #260. @LoyVanBeek once you have implemented your features, we should create a merged feature branch and test/finalize on hardware...

Sound like a good idea. What's the best way to do this? Branch off `kinetic-devel`, then merge both these branches into the new branch? Can be 1 merge commit even

fmessmer commented on cob_script_server/src/simple_script_server/simple_script_server.py:
```diff
@@ -634,29 +644,54 @@ def compose_trajectory(self, component_name, parameter_name):
-				t = math.sqrt(d_max / default_acc)
-			point_time = max(t, 0.4)	# use minimal point_time
+			# d_max: Distance traveled in the move by each joint
+			d_max = numpy.abs(numpy.array(start_pos) - numpy.array(end_pos))
```
this is an array of distances now, not a scalar distance_max anymore...`d_max` name should be adjusted

LoyVanBeek replied:
The same goes for all of them actually, numpy makes it look the same of course, but `default_vel`, `default_acc`, `d_max`, `t1`, `s1`, `t` as all arrays. I think the comment is clear about it and I prefer to not make the name more complicated but keep the expressions short.

fmessmer replied:
I rather meant that `d_max` is not even `max` anymore...it's just an array of joint distances...the array part is ok...

fmessmer reviewed (changes requested):
minor

fmessmer commented:
giving this PR a try - in gazebo simulation:

![Screenshot from 2019-11-12 10-47-19](https://user-images.githubusercontent.com/508006/68660705-fe8ab780-0539-11ea-9739-0a8e40755d0b.png)

explanation - bumps from left to right:
 - bump 1 and 2 - are `cob_command_tools/indigo_dev` behavior for `sss.move("torso","back")` and `sss.move("torso","front")`
 - bump 3 and 4 - are `cob_command_tools/feature/faster_sss_trajectories` behavior for `sss.move("torso","back")` and `sss.move("torso","front")`
 - bump 5 - is `cob_command_tools/feature/faster_sss_trajectories` behavior for `sss.move("torso","back",speed_factor=0.5)`
 - bump 6 - is `cob_command_tools/feature/faster_sss_trajectories` behavior for `sss.move("torso","front",speed_factor=0.25)`

fazit:
 - feature works in principle
 - with `speed_factor=1.0` (default) `follow_joint_trajectory` action aborts with `PATH_TOLERANCE_VIOLATED` - at least for gazebo simulation 
    :point_right: note that `follow_joint_trajectory` controller is implemented in a way that it still continues to execute the motion although the action already aborted

:warning: **this is pr is a significant behavior change when used on a robot** :warning: 
i.e. all robot movements will be much faster as the `default_vel` is significantly lower then the `velocity_limits`
this might - in the least - surprise anybody not knowing about the change and - in the worst - lead to damage to robot and customer
thus, we should implement this speedup as a **opt-in** - disabled by default - to keep behavior as is for all current `sss.move` occurences in any robot/scenario and only enable this explicitly where wanted, i.e. `mbl`
this, can be done e.g. by passing another argument to toggle between using `default_vel` or `urdf_vel` - using `default_vel` as default

@LoyVanBeek

LoyVanBeek commented:
As for the latter (optional `urdf_velocity=True`): I was think the same thing yesterday evening, so I'll implement that.

LoyVanBeek committed:
Optionally use the URDF-derived velocities

LoyVanBeek committed:
Rename d_max to dist

fmessmer committed:
allow setting default_vel and default_acc per joint as list of float/int via yaml

fmessmer committed:
allow to pass default_vel via sss argument

fmessmer committed:
expose arguments to move_rel

fmessmer committed:
return failure on invalid arguments

fmessmer committed:
add vel_limit check

fmessmer commented:
@LoyVanBeek please consider https://github.com/LoyVanBeek/cob_command_tools/pull/1

fmessmer committed:
fixup limit_vel and add debug output

fmessmer committed:
fixup variable name

fmessmer committed:
fixup acc default value

fmessmer committed:
fix error case velocities zero or negative

LoyVanBeek committed:
feature/calculate traj point velocities (#1)

feature/calculate traj point velocities

fmessmer commented on cob_script_server/src/simple_script_server/simple_script_server.py:
```diff
@@ -594,19 +599,71 @@ def compose_trajectory(self, component_name, parameter_name):
+			elif (type(default_vel) is list) and (len(default_vel) == len(start_pos)) and all(((type(item) is float) or (type(item) is int)) for item in default_vel):
+				default_vel = default_vel
+			else:
+				rospy.logerr("parameter %s has wrong format (must be float/int or list of float/int) with proper dimensions.")
```
https://github.com/LoyVanBeek/cob_command_tools/pull/1#discussion_r345760065

should be: `argument 'default_vel' has wrong format...`

LoyVanBeek committed:
Fix undeclared variable

LoyVanBeek committed:
Move all code to determine the desired velocity to a private method

Code to determine this got too big to contain all the various branches and edge cases

fmessmer commented:
@LoyVanBeek 
are you happy with this pr?
I'll try to test and merge today - if it's the case...

LoyVanBeek commented:
I'm OK with the current code, good to test on real HW. 
A next PR would run a PEP-8-ifyer over the code to fix the formatting.

fmessmer committed:
explicit robot_description key

fmessmer committed:
fix log formatting

fmessmer reviewed (approved):
tested on `cob4-21` :heavy_check_mark: 

without explicitly setting any of the new arguments `urdf_vel`, `default_vel`, `speed_factor` the behavior is as before...
